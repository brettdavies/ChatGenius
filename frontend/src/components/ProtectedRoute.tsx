import { useAuth0 } from '@auth0/auth0-react';
import { Navigate, useLocation } from 'react-router-dom';

interface ProtectedRouteProps {
  children: React.ReactNode;
}

const ProtectedRoute = ({ children }: ProtectedRouteProps) => {
  const { isAuthenticated, isLoading } = useAuth0();
  const location = useLocation();

  if (isLoading) {
    return (
      <div className="min-h-screen flex items-center justify-center">
        <div 
          role="status"
          aria-label="Loading"
          data-testid="loading-spinner"
          className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-indigo-500"
        >
          <span className="sr-only">Loading...</span>
        </div>
      </div>
    );
  }

  if (!isAuthenticated) {
    return <Navigate to="/login" state={{ from: location }} replace />;
  }

  return <>{children}</>;
};

export default ProtectedRoute; 