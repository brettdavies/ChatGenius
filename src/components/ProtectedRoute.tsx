import { useAuth0 } from '@auth0/auth0-react';
import { Navigate, useLocation } from 'react-router-dom';

interface ProtectedRouteProps {
  children: React.ReactNode;
}

const ProtectedRoute = ({ children }: ProtectedRouteProps) => {
  const { isAuthenticated, isLoading } = useAuth0();
  const location = useLocation();

  if (isLoading) {
    return (
      <div className="min-h-screen flex items-center justify-center">
        <div className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-indigo-500"></div>
      </div>
    );
  }

  if (!isAuthenticated) {
    return <Navigate to="/login" state={{ from: location }} replace />;
  }

  return <>{children}</>;
};

export default ProtectedRoute; 