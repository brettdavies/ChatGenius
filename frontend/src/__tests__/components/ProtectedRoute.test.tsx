import { render, screen } from '@testing-library/react';
import { useAuth0 } from '@auth0/auth0-react';
import { MemoryRouter, Route, Routes, useLocation } from 'react-router-dom';
import ProtectedRoute from '../../components/ProtectedRoute';

// Mock useAuth0
jest.mock('@auth0/auth0-react');
const mockUseAuth0 = useAuth0 as jest.Mock;

// Mock component to capture location state
const LocationDisplay = () => {
  const location = useLocation();
  return <div data-testid="location-state">{JSON.stringify(location.state)}</div>;
};

describe('ProtectedRoute', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('snapshots', () => {
    it('matches snapshot when loading', () => {
      mockUseAuth0.mockReturnValue({
        isLoading: true,
        isAuthenticated: false
      });

      const { container } = render(
        <MemoryRouter>
          <ProtectedRoute>
            <div>Protected Content</div>
          </ProtectedRoute>
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when authenticated', () => {
      mockUseAuth0.mockReturnValue({
        isLoading: false,
        isAuthenticated: true
      });

      const { container } = render(
        <MemoryRouter>
          <ProtectedRoute>
            <div>Protected Content</div>
          </ProtectedRoute>
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when not authenticated', () => {
      mockUseAuth0.mockReturnValue({
        isLoading: false,
        isAuthenticated: false
      });

      const { container } = render(
        <MemoryRouter initialEntries={['/protected']}>
          <Routes>
            <Route path="/login" element={<LocationDisplay />} />
            <Route
              path="/protected"
              element={
                <ProtectedRoute>
                  <div>Protected Content</div>
                </ProtectedRoute>
              }
            />
          </Routes>
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('shows loading spinner when authentication is loading', () => {
    mockUseAuth0.mockReturnValue({
      isLoading: true,
      isAuthenticated: false
    });

    render(
      <MemoryRouter>
        <ProtectedRoute>
          <div>Protected Content</div>
        </ProtectedRoute>
      </MemoryRouter>
    );

    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('redirects to login when user is not authenticated', () => {
    mockUseAuth0.mockReturnValue({
      isLoading: false,
      isAuthenticated: false
    });

    const testPath = '/protected';

    render(
      <MemoryRouter initialEntries={[testPath]}>
        <Routes>
          <Route path="/login" element={<LocationDisplay />} />
          <Route
            path={testPath}
            element={
              <ProtectedRoute>
                <div>Protected Content</div>
              </ProtectedRoute>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    // Verify we're redirected and the location state is preserved
    const locationState = screen.getByTestId('location-state');
    expect(locationState).toHaveTextContent(testPath);
  });

  it('renders children when user is authenticated', () => {
    mockUseAuth0.mockReturnValue({
      isLoading: false,
      isAuthenticated: true
    });

    render(
      <MemoryRouter>
        <ProtectedRoute>
          <div data-testid="protected-content">Protected Content</div>
        </ProtectedRoute>
      </MemoryRouter>
    );

    expect(screen.getByTestId('protected-content')).toBeInTheDocument();
    expect(screen.getByText('Protected Content')).toBeInTheDocument();
  });

  it('preserves location state when redirecting', () => {
    mockUseAuth0.mockReturnValue({
      isLoading: false,
      isAuthenticated: false
    });

    const testPath = '/protected';
    const testState = { someData: 'test-data' };

    render(
      <MemoryRouter initialEntries={[{ pathname: testPath, state: testState }]}>
        <Routes>
          <Route path="/login" element={<LocationDisplay />} />
          <Route
            path={testPath}
            element={
              <ProtectedRoute>
                <div>Protected Content</div>
              </ProtectedRoute>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    // Verify the location state includes both the 'from' path and original state
    const locationState = screen.getByTestId('location-state');
    expect(locationState).toHaveTextContent(testPath);
    expect(locationState).toHaveTextContent('test-data');
  });
}); 