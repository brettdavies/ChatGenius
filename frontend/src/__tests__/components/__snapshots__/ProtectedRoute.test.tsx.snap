// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute snapshots matches snapshot when authenticated 1`] = `
<div>
  <div>
    Protected Content
  </div>
</div>
`;

exports[`ProtectedRoute snapshots matches snapshot when loading 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center"
  >
    <div
      aria-label="Loading"
      class="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-indigo-500"
      data-testid="loading-spinner"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`ProtectedRoute snapshots matches snapshot when not authenticated 1`] = `
<div>
  <div
    data-testid="location-state"
  >
    {"from":{"pathname":"/protected","search":"","hash":"","state":null,"key":"default"}}
  </div>
</div>
`;
